_posts/2022-04-22-switch-to-java-11.adoc

```diff
@@ -0,0 +1,138 @@
+author: vjuranek
+---
+
+As you probably noticed, we have started work on Debezium 2.0.
+One of https://issues.redhat.com/browse/DBZ-3899[the planned changes] for the 2.0 release is https://issues.redhat.com/browse/DBZ-4949[to switch to Java 11 as a baseline].
+While some Java implementations still supports Java 8 
+(e.g. OpenJDK will be https://access.redhat.com/articles/1299013[supported by Red Hat] till 2026), 
+free Oracle JDK 8 already https://www.oracle.com/java/technologies/java-se-support-roadmap.html[has reached end of life]. 
+So it's time to start thinking about using a more recent JDK, and the new major Debezium release is a natural milestone when to do the switch.
+
++++<!-- more -->+++
+
+Since early releases of Debezium 2.0, hopefully from the very first one `Alpha1`, Debezium bits will be compiled to Java 11 byte code.
+Therefore, Java 11 will be required to run Debezium in the next major update.
+Also, if you use any of the Debezium bits in your project, you will have to switch to Java 11.
+
+But wait, what does Java 11/17 in the title mean?
+Is it there just to scare you or we are going to actually switch to Java 17?
+
+<dramatic pause here>
+
+No, we don't want to scare you.
+We are actually planning to switch to Java 17, but only for the test suite.
+This will allow us to use more recent Java features in the tests, while users still can stay on Java 11.
+
+== Implementation
+
+Setting different byte code level for code and tests is pretty easy with Maven, 
+you just need to set following system properties:
+
+[source, xml]
+----
+<maven.compiler.release>11</maven.compiler.release>
+<maven.compiler.testRelease>17</maven.compiler.testRelease>
+----
+
+After switching to Java 11, checkstyle plugin and impsort plugin (plugin which takes care about proper order of imports) started to fail.
+However, bumping their versions to the latest releases has solved all the issues.
+
+This was the easy part.
+The most difficult part was the Cassandra connector.
+
+=== Cassandra connector
+
+Cassandra connector now provides support for Cassandra 3 as well as for Cassandra 4.
+Cassandra 4 https://cassandra.apache.org/doc/4.0/cassandra/new/java11.html[works with Java 11], 
+but running Cassandra 3 with Java 11 is not possible (or at least requires some hacking).
+Current test implementation doesn't run Cassandra in a container as we do in tests for other DB connectors, but runs Cassandra in embedded mode.
+Therefore if you want to run the tests with Java 11 (or 17), tests for Cassandra 3 module would fail.
+Obvious solution is to run Cassandra in a container with Java 8.
+This sounds good, but this approach has one pitfall.
+The Cassandra connector needs access to Cassandra log files as it reads the dedicated log file to obtain CDC events, so the tests need to access Cassandra files in the container.
+This can be solved quite easily using a temporary directory, .one in e.g `target`, giving access to all users and moutable within the container.
```

Comment:
```suggestion
This can be solved quite easily using a temporary directory, for instance within the `target` directory, and mount it as a volume into the container running Cassandra.
```